### PR TITLE
Switch cache to ioredis

### PR DESCRIPTION
**Pull request by karcass**
## 📚 Purpose
Address issues created by somewhat janky old-redis client combined with bluebird. use same ioredis client from lacuna-tech/lacuna-core.

### CODE REVIEW

**Comment by karcass** [comment]
Should this be build-able?

File: packages/mds-agency-cache/index.ts [hunk 1]
```diff
@@ -139,22 +94,20 @@ async function hread(suffix: string, device_id: UUID): Promise<CachedItem> {
...
   const start = now()
-  const client = await getClient()
   const [pt1, pt2] = bbox
   const points = bbox.map(pt => {
     return { lat: pt[0], lng: pt[1] }
   })
   const [lng, lat] = [(pt1[0] + pt2[0]) / 2, (pt1[1] + pt2[1]) / 2]
   const radius = routeDistance(points)
-  const events = client.georadiusAsync(decorateKey('locations'), lng, lat, radius, 'm')
+  const events = client.georadius(decorateKey('locations'), lng, lat, radius, 'm')
```
Comment (drtyh2o) [inline packages/mds-agency-cache/index.ts:110]: Should this be `await`-ed? `events` is just going to be a promise here, I believe, so it won't be included in the `timeElapsed`.

File: packages/mds-cache/redis/index.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        client.disconnect()
+      }
+      client = null
+    },
+    multi: async () => {
+      return safelyExec(theClient => {
+        return theClient.multi()
+      })
+    },
+    get: async (key: KeyType) => {
```
Comment (drtyh2o) [inline packages/mds-cache/redis/index.ts:41]: What was the motivation for removing the _Async_ suffix convention? `get` vs `getAsync`, etc.?

**Review by drtyh2o — APPROVED** [review]
A few minor nits and questions. Nothing blocking. I assume the unit tests confirm that all of the async/await are in the proper places.

File: packages/mds-cache/redis/index.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  let client: Nullable<Redis.Redis> = null
+
+  /**
+   * If the client is defined, the closure is called, otherwise throws an error
+   * @param closure called with a Redis client, returns the result
+   * @returns same as what the closure returns
+   * @throws ClientDisconnectedError
+   */
+
+  const safelyExec = async <T>(closure: (theClient: Redis.Redis) => T) => {
```
[the remark this answers is not in the document]
Comment (karcass) [reply packages/mds-cache/redis/index.ts:19]: uh ... ask @avatarneil ?

File: packages/mds-agency-cache/index.ts (hunk 1 above)
Comment (karcass) [reply packages/mds-agency-cache/index.ts:110]: Yep, this looks like it was a mistake from Way Back.  Probably should have types on e.g. `events`.

File: packages/mds-cache/redis/index.ts (hunk 2 above)
Comment (karcass) [reply packages/mds-cache/redis/index.ts:41]: It was entirely an artifact of Bluebird.

File: packages/mds-cache/redis/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (avatarneil) [reply packages/mds-cache/redis/index.ts:19]: That's a @nessur question :)

**Comment by janedotx** [comment]
> Should this be build-able? As in, is the Jenkins failure significant?

I'm seeing the same error when I check out your branch and run it locally. It's not there when I run `develop`.

**Comment by drtyh2o** [comment]
> > Should this be build-able? As in, is the Jenkins failure significant?
> 
> I'm seeing the same error when I check out your branch and run it locally. It's not there when I run `develop`.

Test timeout? ... could possibly be a missing `await` somewhere. I think jest spins up tests in separate processes so if you don't await a promise then the process doesn't exit and causes a timeout. Does you get the same or different result if you run tests with the `--runInBand` option?

**Comment by janedotx** [comment]
> > > Should this be build-able? As in, is the Jenkins failure significant?
> > 
> > 
> > I'm seeing the same error when I check out your branch and run it locally. It's not there when I run `develop`.
> 
> Test timeout? ... could possibly be a missing `await` somewhere. I think jest spins up tests in separate processes so if you don't await a promise then the process doesn't exit and causes a timeout. Do you get the same or different result if you run tests with the `--runInBand` option?

`mds-daily` is using `ts-mocha` still, judging by its `package.json`, and I think those tests are running serially.

**Comment by janedotx** [comment]
```
(node:77982) UnhandledPromiseRejectionWarning: ClientDisconnectedError: Client is not connected! Have you tried initializing the client with .initialize()?
    at new ClientDisconnectedError (/Users/jane/work/mds-core/packages/mds-utils/exceptions/exceptions.ts:91:5)
    at safelyExec (/Users/jane/work/mds-core/packages/mds-cache/redis/index.ts:23:11)
    at Object.flushdb (/Users/jane/work/mds-core/packages/mds-cache/redis/index.ts:67:14)
    at reset (/Users/jane/work/mds-core/packages/mds-agency-cache/index.ts:487:16)
    at Object.initialize (/Users/jane/work/mds-core/packages/mds-agency-cache/index.ts:493:9)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 1)
```

I see this error, and it's definitely in `tests/daily.spec.ts`. Probably something is wrong with `cache.initialize()` in the `before` block.

**Review by janedotx — CHANGES_REQUESTED** [review]
mds-daily tests need to pass

**Comment by janedotx** [comment]
Agency is having problems too:
```
  1) "before all" hook in "{root}":
     ClientDisconnectedError: Client is not connected! Have you tried initializing the client with .initialize()?
      at new ClientDisconnectedError (/Users/jane/work/mds-core/packages/mds-utils/exceptions/exceptions.ts:91:5)
      at safelyExec (/Users/jane/work/mds-core/packages/mds-cache/redis/index.ts:23:11)
      at Object.flushdb (/Users/jane/work/mds-core/packages/mds-cache/redis/index.ts:67:14)
      at reset (/Users/jane/work/mds-core/packages/mds-agency-cache/index.ts:487:16)
      at Object.initialize (/Users/jane/work/mds-core/packages/mds-agency-cache/index.ts:493:9)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
      at async Promise.all (index 1)
      at Context.<anonymous> (tests/integration-tests.spec.ts:133:3)

```

**Review by avatarneil — CHANGES_REQUESTED** [review]
I think mds-daily is not the only failing package... Agency tests are failing for me too. It looks like the cache is never actually being initialized, you need to be calling `client.initialize()` _somewhere_... When I updated `packages/mds-agency-cache/index.ts` `initialize()` to call `await client.initialize()` instead of `await getClient()` (a call which does literally nothing), agency tests actually ran, but there were 15 failures. Similar behavior in daily.

File: packages/mds-agency-cache/index.ts [hunk 4]
```diff
@@ -45,64 +45,19 @@ const { env } = process
...
-    hgetallAsync: (arg1: string) => Promise<{ [key: string]: string }>
-    execAsync: () => Promise<object[]>
-  }
-}
-
-bluebird.promisifyAll(redis.RedisClient.prototype)
-bluebird.promisifyAll(redis.Multi.prototype)
-
-let cachedClient: redis.RedisClient | null
+const client = RedisCache()
```
Comment (janedotx) [inline packages/mds-agency-cache/index.ts:48]: I think this should be something more like `RedisCache().initialize()`.

**Comment by janedotx** [comment]
> Should this be build-able? As in, is the Jenkins failure significant?

I'm very curious. Did this build locally for you?

File: packages/mds-cache/redis/index.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      return safelyExec(theClient => {
+        return theClient.get(key)
+      })
+    },
+    set: async (key: KeyType, val: ValueType) => {
+      return safelyExec(theClient => {
+        return theClient.set(key, val)
+      })
+    },
+    expireAt: async (key: KeyType, time: Timestamp) => {
```
[the remark this answers is not in the document]
Comment (karcass) [reply packages/mds-cache/redis/index.ts:51]: > > Should this be build-able? As in, is the Jenkins failure significant?
> 
> I'm very curious. Did this build locally for you?

Everything passed tsc.  The only failure I saw was `mds-daily`.  I'm wondering if I missed the `agency` stuff when scrolling back through the zillions of lines of output.

**Comment by karcass** [comment]
Okay, I'm a dumbass.  It had been so long since I ran tests, I thought it ran all the tests, and didn't stop on the first failure.

PR changed to draft.

**Comment by avatarneil** [comment]
Note, though tests are passing, we need to add explicit cache initialization at server start time before this can be merged. Otherwise, every deployment will be 💩

File: packages/mds-agency-cache/index.ts (hunk 4 above)
Comment (karcass) [reply packages/mds-agency-cache/index.ts:46]: See changes

**Review by nessur — APPROVED** [review]

**Review by janedotx — APPROVED** [review]

**Comment by drtyh2o** [comment]
Does `ioredis` support the streaming APIs used by `mds-stream` too? Might be nice to have a similar cleanup to the redis/bluebird dependencies there as well.

File: packages/mds-agency-cache/index.ts [hunk 6]
```diff
@@ -453,6 +387,7 @@ async function readDevicesStatus(query: {
     Object.assign(deviceStatusMap[item.device_id], item)
   })
   const values = Object.values(deviceStatusMap)
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (drtyh2o) [inline packages/mds-agency-cache/index.ts:390]: I would rather leave these sorts of warnings in place to identify opportunities to tighten up the typings.

**Review by avatarneil — APPROVED** [review]

**Pull request merged**